File: 1Traveler.py
import sys
if __name__ == '__main__':
    # read input
    n = int(sys.stdin.readline())
    # save the balls according to its color
    colorEnds = {}
    for _ in range(n):
        ballPos, ballColor = sys.stdin.readline().split()
        # convert string to int
        # we just need to find the left and right end of each color
        # because the optimal solution must gurantee that
        # each color starts end ends at its two ends.
        # the third element is step length
        # because we have to collect them all
        ballPos = int(ballPos)
        ballColor = int(ballColor)

        # save the balls according to its color
        if not colorEnds.__contains__(ballColor):
            # if the color does not in the dict
            # create a new key setting the two ends as the same
            colorEnds[ballColor] = [ballPos, ballPos]
        else:
            colorEnds[ballColor][0] = min(colorEnds[ballColor][0], ballPos)
            colorEnds[ballColor][1] = max(colorEnds[ballColor][1], ballPos)
    # print(colorEnds)

    # because color id does not start from 1
    # so we firstly collect all the colors and sort
    colors = list(colorEnds.keys())
    colors.sort()
    colors.append(0)
    # we use DP to solve this problem
    # dp[i] means the min time 
    # to collect the balls of the first i colors
    # Note that each dp[i] has four elements,
    # idx 0: num of total steps if start collecting from the left end
    # idx 1: ending position if start collecting from the left end
    # idx 2: num of total steps if start collecting from the right end
    # idx 3: ending position if start collecting from the right end

    # add an [0, 0] to go back to zero
    colorEnds[0] = [0, 0]
    print(colorEnds)

    numOfColors = len(colors)
    # print("num of colors:", numOfColors)
    dp = [[0,0,0,0]]

    # remind that the color id does not start from 1
    # but we have to collect from the smallest color ID
    for i in range(1, numOfColors+1):
        currentDP = [0,0,0,0]

        colorIdx = colors[i-1]
        # leftEnd of current color
        leftEnd = colorEnds[colorIdx][0]
        # rightEnd of current color
        rightEnd = colorEnds[colorIdx][1]
        # stepLen is collecting length
        stepLen = rightEnd - leftEnd
        # print("colorIdx: ", colorIdx, " leftEnd: ", leftEnd, " rightEnd: ", rightEnd, " len: ", stepLen)

        # if we start from the left end
        # steps continuing from the left end of the last step
        lNumSteps = dp[i-1][0] + stepLen + abs(dp[i-1][1] - rightEnd)
        # steps continuing from the right end of the last step
        rNumSteps = dp[i-1][2] + stepLen + abs(dp[i-1][3] - rightEnd)
        # print("lNumSteps1: ", lNumSteps, " rNumSteps1: ", rNumSteps)
        if lNumSteps < rNumSteps:
            currentDP[0] = lNumSteps
        else:
            currentDP[0] = rNumSteps
        # start from right so end at left
        currentDP[1] = leftEnd

        # if we start from the right end
        # steps continuing from the left end of the last step
        lNumSteps = dp[i-1][0] + stepLen + abs(dp[i-1][1] - leftEnd)
        # steps continuing from the right end of the last step
        rNumSteps = dp[i-1][2] + stepLen + abs(dp[i-1][3] - leftEnd)
        # print("lNumSteps2: ", lNumSteps, " rNumSteps2: ", rNumSteps)
        if lNumSteps < rNumSteps:
            currentDP[2] = lNumSteps
        else:
            currentDP[2] = rNumSteps
        # start from left so end at right
        currentDP[3] = rightEnd
        # print(currentDP)
        dp.append(currentDP)
        # print(dp[-1])

    # do not forget to go back to original position
    minSteps = dp[numOfColors][0]
    print(minSteps)
    # print(dp)

File: 1Traveler/1Traveler.cpp
#include <iostream>
#include <map>
#include <vector>

using namespace std;

class Traveler
{
public:
    // auto sorted
    map<long, vector<long>> colorEnds;
    // num of balls
    long n = 0;
    vector<vector<long>> dp = { {0,0,0,0} };

    void ReadInput();
    void OutputColorEnds();
    void DP();
};

void Traveler::ReadInput()
{
    cin >> n;
    // N <= 2*10^5
    long ballPos, ballColor;
    for (long i = 0; i < n; i++)
    {
        cin >> ballPos >> ballColor;
        // we just need to find the leftand right end of each color
        // because the optimal solution must gurantee that
        // each color starts end ends at its two ends.
        // the third element is step length
        // because we have to collect them all
        if (colorEnds.count(ballColor) == 0)
        {
            colorEnds.insert({ ballColor, {ballPos, ballPos} });
        }
        else
        {
            vector<long> ends = colorEnds[ballColor];
            ends[0] = min(ends[0], ballPos);
            ends[1] = max(ends[1], ballPos);
            colorEnds[ballColor] = ends;
        }
    }
}

void Traveler::OutputColorEnds()
{
    map<long, vector<long>>::const_iterator it;
    for (it = colorEnds.begin(); it != colorEnds.end(); ++it) {
        cout << "Color: " << it->first << " left: " << it->second[0] << " right: " << it->second[1] << endl;
    }
}

void Traveler::DP()
{
    // for each color
    map<long, vector<long>>::const_iterator it;
    long i = 1;
    for (it = colorEnds.begin(); it != colorEnds.end(); ++it) {
        vector<long> currentDP = {0,0,0,0};
        // leftEnd of current color
        long leftEnd = it->second[0];
        // rightEnd of current color
        long rightEnd = it->second[1];
        // stepLen is collecting length
        long stepLen = rightEnd - leftEnd;
        
        // if we start from the left end
        // steps continuing from the left end of the last step
        long lNumSteps = dp[i - 1][0] + stepLen + abs(dp[i - 1][1] - rightEnd);
        // steps continuing from the right end of the last step
        long rNumSteps = dp[i - 1][2] + stepLen + abs(dp[i - 1][3] - rightEnd);
        currentDP[0] = min(lNumSteps, rNumSteps);
        // start from right so end at left
        currentDP[1] = leftEnd;

        // if we start from the right end
        // steps continuing from the left end of the last step
        lNumSteps = dp[i - 1][0] + stepLen + abs(dp[i - 1][1] - leftEnd);
        // steps continuing from the right end of the last step
        rNumSteps = dp[i - 1][2] + stepLen + abs(dp[i - 1][3] - leftEnd);
        currentDP[2] = min(lNumSteps, rNumSteps);
        // start from right so end at left
        currentDP[3] = rightEnd;
        // cout << currentDP[0] << " " << currentDP[1] << " " << currentDP[2] << " " << currentDP[3] << endl;
        dp.push_back(currentDP);
        i++;
    }

    vector<long> currentDP = { 0,0,0,0 };
    long leftEnd = 0;
    // rightEnd of current color
    long rightEnd = 0;
    // stepLen is collecting length
    long stepLen = rightEnd - leftEnd;

    // if we start from the left end
    // steps continuing from the left end of the last step
    long lNumSteps = dp[i - 1][0] + stepLen + abs(dp[i - 1][1] - rightEnd);
    // steps continuing from the right end of the last step
    long rNumSteps = dp[i - 1][2] + stepLen + abs(dp[i - 1][3] - rightEnd);
    currentDP[0] = min(lNumSteps, rNumSteps);
    // start from right so end at left
    currentDP[1] = leftEnd;

    // if we start from the right end
    // steps continuing from the left end of the last step
    lNumSteps = dp[i - 1][0] + stepLen + abs(dp[i - 1][1] - leftEnd);
    // steps continuing from the right end of the last step
    rNumSteps = dp[i - 1][2] + stepLen + abs(dp[i - 1][3] - leftEnd);
    currentDP[2] = min(lNumSteps, rNumSteps);
    // start from right so end at left
    currentDP[3] = rightEnd;

    dp.push_back(currentDP);
    i++;
    // cout << currentDP[0] << " " << currentDP[1] << " " << currentDP[2] << " " << currentDP[3] << endl;
    cout << dp[i-1][0] << endl;
}

int main()
{
    Traveler traveler;
    // # read input
    traveler.ReadInput();
    // traveler.OutputColorEnds();
    traveler.DP();
    return 0;
} 